xds/internal/xdsclient/transport/ads/ads_stream.go

```diff
@@ -0,0 +1,780 @@
+// state is `watchStateStarted`.
+func (s *Stream) sendNew(stream transport.StreamingCall, typ xdsresource.Type) error {
+	s.logger.Infof("easwars: sendNew for resource type %q", typ.TypeName())
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Ig there's no stream yet, skip the request. This request will be resent
+	// when a new stream is created. If no stream is created, the watcher will
+	// timeout (same as server not sending response back).
+	if stream == nil {
+		return nil
+	}
+
+	// If local processing of the most recently received response is not yet
+	// complete, i.e. fc.pending == true, queue this write and return early.
+	// This allows us to batch writes for requests which are generated as part
+	// of local processing of a received response.
+	state := s.resourceTypeState[typ]
+	if s.fc.pending.Load() {
+		select {
+		case state.bufferedRequests <- struct{}{}:
+		default:
+		}
+		s.logger.Infof("easwars: buffered type %q", typ.TypeName())
+		return nil
+	}
+
+	names := resourceNames(state.subscribedResources)
+	s.logger.Infof("easwars: sendNew for resource type %q, names: %v", typ.TypeName(), names)
+	if err := s.sendMessageLocked(stream, names, typ.TypeURL(), state.version, state.nonce, nil); err != nil {
+		return err
+
+	}
+	select {
+	case <-state.bufferedRequests:
+	default:
+	}
+	s.startWatchTimersLocked(typ, names)
+	return nil
+}
+
+// sendExisting sends out discovery requests for existing resources when
+// recovering from a broken stream.
+//
+// The stream argument is guaranteed to be non-nil.
+func (s *Stream) sendExisting(stream transport.StreamingCall) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for typ, state := range s.resourceTypeState {
+		// Reset only the nonces map when the stream restarts.
+		//
+		// xDS spec says the following. See section:
+		// https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-resource-type-instance-version
+		//
+		// Note that the version for a resource type is not a property of an
+		// individual xDS stream but rather a property of the resources
+		// themselves. If the stream becomes broken and the client creates a new
+		// stream, the client’s initial request on the new stream should
+		// indicate the most recent version seen by the client on the previous
+		// stream
+		state.nonce = ""
+
+		if len(state.subscribedResources) == 0 {
+			continue
+		}
+
+		names := resourceNames(state.subscribedResources)
+		if s.logger.V(2) {
+			s.logger.Infof("Re-requesting resources %v of type %q, as the stream has been recreated", names, typ.TypeURL())
+		}
+		if err := s.sendMessageLocked(stream, names, typ.TypeURL(), state.version, state.nonce, nil); err != nil {
+			return err
+		}
+		select {
+		case <-state.bufferedRequests:
+		default:
+		}
+		s.startWatchTimersLocked(typ, names)
+	}
+	return nil
+}
+
+// sendBuffered sends out discovery requests for resources that were buffered
+// (when they were subscribed to) because of lack of flow control quota.
```

Comment:
I thought state.bufferedRequests also corresponds to batching the writes while processing previous resource?

Reply:
It's not really "quota" then is it, it's just it batches writes which is binary on or off, quota to me refers to numerical flow control bytes.

Reply:
How about now?